test/e2e/dataplane/service_discovery.go

```diff
@@ -42,4 +50,69 @@ func RunServiceDiscoveryTest(f *framework.Framework) {
 
 	Expect(err).NotTo(HaveOccurred())
 	Expect(stdout).To(ContainSubstring("Welcome to nginx!"))
+
+	kubeDnsService, _ := f.ClusterClients[framework.ClusterB].CoreV1().Services("kube-system").Get("kube-dns", metav1.GetOptions{})
+	kubeDnsServiceIP := kubeDnsService.Spec.ClusterIP
+
+	cmd = []string{"dig", "@" + kubeDnsServiceIP, "nginx-demo." + f.Namespace + ".svc.cluster2.local", "+short"}
+	stdout, _, err = f.ExecWithOptions(framework.ExecOptions{
+		Command:       cmd,
+		Namespace:     f.Namespace,
+		PodName:       netshootPodList.Items[0].Name,
+		ContainerName: netshootPodList.Items[0].Spec.Containers[0].Name,
+		CaptureStdout: true,
+		CaptureStderr: true,
+	}, framework.ClusterB)
+
+	Expect(err).NotTo(HaveOccurred())
+	Expect(stdout).To(ContainSubstring(nginxServiceClusterCIP))
+	f.DeleteService(framework.ClusterC, "nginx-demo")
+	f.AwaitForMcsCrdDelete(framework.ClusterB, "nginx-demo")
+	stdout, _, err = f.ExecWithOptions(framework.ExecOptions{
+		Command:       cmd,
+		Namespace:     f.Namespace,
+		PodName:       netshootPodList.Items[0].Name,
+		ContainerName: netshootPodList.Items[0].Spec.Containers[0].Name,
+		CaptureStdout: true,
+		CaptureStderr: true,
+	}, framework.ClusterB)
+
+	Expect(err).NotTo(HaveOccurred())
+	Expect(stdout).NotTo(ContainSubstring(nginxServiceClusterCIP))
```

Comment:
It feels like `AwaitForMcsCrdDelete` is a bit whitebox, also it seems it's possible the core DNS plugin may not yet have observed the deletion in which case the test would fail.  I suggest to enclose the `ExecWithOptions` call in `AwaitUntil`, eg

```
By(fmt.Sprintf("Executing %q to verify Nginx service is no longer discoverable", strings.Join(cmd, " "))
AwaitUntil("verify service IP is not discoverable", func() (interface{}, error) {
    stdout, _, err := f.ExecWithOptions(...)
    if err != nil {
        return nil, err
    }
    return stdout, nil
}, func(result interface{}) (bool, error) {
    if strings.Contains(result.(string), nginxServiceClusterCIP) {
        return false, nil
    }
    return true, nil
})
```

Reply:
Done.